src/payment/strategies/googlepay/googlepay-payment-strategy.ts

```diff
@@ -0,0 +1,371 @@
+
+        this._walletButton = undefined;
+
+        return this._googlePayInitializer.teardown()
+            .then(() => super.deinitialize(options));
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        return this._getPayment()
+            .catch((error: MissingDataError) => {
+                if (error.subtype === MissingDataErrorType.MissingPayment) {
+                    return this._displayWallet()
+                        .then(() => this._getPayment());
+                }
+
+                throw error;
+            })
+            .then(payment => {
+                return this._createOrder(payment, payload.useStoreCredit, options);
+            });
+    }
+
+    createButton(): HTMLElement {
+        return this._googlePaymentsClient.createButton({
+            buttonColor: ButtonColor.default,
+            buttonType: ButtonType.short,
+            onClick: this._handleWalletButtonClick,
+        });
+    }
+
+    private _configureWallet(): Promise<void> {
+        if (!this._methodId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(this._methodId))
+            .then(state => {
+                const paymentMethod = state.paymentMethods.getPaymentMethod(this._methodId);
+                const storeConfig = state.config.getStoreConfig();
+                const checkout = state.checkout.getCheckout();
+                const hasShippingAddress = !!state.shippingAddress.getShippingAddress();
+
+                if (!paymentMethod) {
+                    throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+                }
+
+                if (!storeConfig) {
+                    throw new MissingDataError(MissingDataErrorType.MissingCheckoutConfig);
+                }
+
+                if (!checkout) {
+                    throw new MissingDataError(MissingDataErrorType.MissingCheckout);
+                }
+
+                this._paymentMethod = paymentMethod;
+                const testMode = paymentMethod.config.testMode;
+
+                return Promise.all([
+                    this._googlePayScriptLoader.load(),
+                    this._googlePayInitializer.initialize(checkout, paymentMethod, hasShippingAddress),
+                ])
+                    .then(([googlePay, googlePayPaymentDataRequest]) => {
+                        this._googlePaymentsClient = this._getGooglePaymentsClient(googlePay, testMode);
+                        this._googlePaymentDataRequest = googlePayPaymentDataRequest;
+                    })
+                    .catch((error: Error) => {
+                        this._handleError(error);
+                    });
+            });
+    }
+
+    private _displayWallet(): Promise<InternalCheckoutSelectors> {
+        return new Promise((resolve, reject) => {
```

Comment:
Is there a reason why this promise never gets resolved? I can see `reject` gets called at the bottom of this method. But I can't see `resolve` getting called anywhere. Also, is there a reason why you need to call `new Promise()`? It seems `isReadyToPay` already returns a promise so you should be able to just chain it with other asynchronous calls and return the resulting promise.

Reply:
I'm gonna change to this: